plugins/3scale-provider/src/providers/ThreeScaleApiEntityProvider.ts

```diff
@@ -0,0 +1,229 @@
+        providerConfig,
+        options.logger,
+        taskRunner,
+      );
+    });
+  }
+
+  private constructor(
+    config: ThreeScaleConfig,
+    logger: Logger,
+    taskRunner: TaskRunner,
+  ) {
+    this.env = config.id;
+    this.baseUrl = config.baseUrl;
+    this.accessToken = config.accessToken;
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+
+    this.scheduleFn = this.createScheduleFn(taskRunner);
+  }
+
+  private createScheduleFn(taskRunner: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:run`;
+      return taskRunner.run({
+        id: taskId,
+        fn: async () => {
+          try {
+            await this.run();
+          } catch (error) {
+            this.logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  getProviderName(): string {
+    return `ThreeScaleApiEntityProvider:${this.env}`;
+  }
+
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn();
+  }
+
+  async run(): Promise<void> {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    this.logger.info(`Discovering ApiEntities from 3scale ${this.baseUrl}`);
+
+    const entities: Entity[] = [];
+
+    let page: number = 0;
+    let services: Services;
+    let apiDocs: APIDocs;
+    let fetchServices: boolean = true;
+    while (fetchServices) {
+      services = await listServices(
+        this.baseUrl,
+        this.accessToken,
+        page,
+        ThreeScaleApiEntityProvider.SERVICES_FETCH_SIZE,
+      );
+      apiDocs = await listApiDocs(this.baseUrl, this.accessToken);
+      for (const element of services.services) {
+        const service = element;
+        this.logger.debug(`Find service ${service.service.name}`);
+
+        // Trying to find the API Doc for the service.
+        const apiDoc = apiDocs.api_docs.find(obj => {
+          return obj.api_doc.service_id === service.service.id;
+        });
+        const proxy = await getProxyConfig(
+          this.baseUrl,
+          this.accessToken,
+          service.service.id,
+        );
+        if (apiDoc !== null) {
+          const apiEntity: ApiEntity = this.buildApiEntityFromService(
+            service,
+            apiDoc!,
+            proxy,
+          );
+          entities.push(apiEntity);
+          this.logger.info(`Discovered ApiEntity ${service.service.name}`);
+        }
+      }
+
+      if (
+        services.services.length <
+        ThreeScaleApiEntityProvider.SERVICES_FETCH_SIZE
+      ) {
+        fetchServices = false;
+      }
+      page++;
+    }
+
+    this.logger.info(`Applying the mutation with ${entities.length} entities`);
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: entities.map(entity => ({
+        entity,
+        locationKey: 'ThreeScaleApiEntityProvider',
```

Comment:
This should at least match the `getProviderName`. In case you have multiple instances of this provider (for different environments for example) and more than one of them emit an entity with the same kind/name/namespace, it will cause conflicts in catalog database.

See: https://backstage.io/docs/features/software-catalog/external-integrations#provider-mutations